File: R/waic_loo.R
#' Compute the WAIC for the model
#'
#' @param mod A BBT model
#'
#' @return the waic, as computed by the loo package
#' @export
#'
#' @examples
#' \donttest{
#' ss <- ll[1:80, 1:6] # first 20 data sets and 5 algorithms
#'
#' m1 <- bbtcomp(ss)
#' get_waic(m1)
#' }
get_waic <- function(mod) {
  log_lik = posterior::as_draws_matrix(mod$model$draws('log_lik'))
  return(loo::waic(log_lik))
}


#' Compute the loo for the model
#'
#' @param mod A BBT model
#'
#' @return The loo, as computed by the loo package
#' @export
#'
#' @examples
#' \donttest{
#' ss <- ll[1:80, 1:6] # first 20 datasets and 5 algorithms
#'
#' m1 <- bbtcomp(ss)
#' get_loo(m1)
#' }
get_loo <- function(mod) return(mod$model$loo())

File: bbtcomp_python/src/bbtcomp/_waic_loo.py
"""WAIC and PSIS-LOO model-fit criteria, via arviz."""

from __future__ import annotations

from ._mcmc import BBTModel


def _idata(mod: BBTModel):
    import arviz as az
    return az.from_cmdstanpy(posterior=mod.model, log_likelihood="log_lik")


def get_waic(mod: BBTModel):
    """Compute the WAIC for a fitted BBT model (via ``arviz.waic``)."""
    import arviz as az
    return az.waic(_idata(mod))


def get_loo(mod: BBTModel):
    """Compute the PSIS-LOO for a fitted BBT model (via ``arviz.loo``)."""
    import arviz as az
    return az.loo(_idata(mod))
